### PR TITLE
Dynamically updating rate limits

### PR DESCRIPTION
**Pull request by Edwin9292**

### CODE REVIEW

File: bucket4j-spring-boot-starter-context/pom.xml [hunk 1]
```diff
@@ -41,7 +41,15 @@
...
 			<optional>true</optional>
 		</dependency>
-	</dependencies>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.tomcat.embed</groupId>
+            <artifactId>tomcat-embed-core</artifactId>
```
Comment (MarcGiffing) [inline bucket4j-spring-boot-starter-context/pom.xml:50]: What is the reason for this library. Not everyone wants to use a the Tomcat.

File: examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java [hunk 2]
```diff
@@ -1,20 +1,146 @@
...
-	
+
 	@GetMapping("world")
 	public ResponseEntity<String> world() {
 		return ResponseEntity.ok("Hello World");
 	}
-	
+
+	@PostMapping("filters/{filterId}")
+	public ResponseEntity updateConfig(
```
Comment (MarcGiffing) [inline examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java:50]: I havn't fully reviewed the PR but can you move this configuration to the bucket4j-spring-boot-starter project? So the update mechanism is available for the users wihtout the necessity to know the required validation.

File: bucket4j-spring-boot-starter-context/pom.xml (hunk 1 above)
Comment (Edwin9292) [reply bucket4j-spring-boot-starter-context/pom.xml:50]: I believe this was needed for making the predicates work, but i'll check tomorrow to be sure and will get back to you.

File: examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java (hunk 2 above)
Comment (Edwin9292) [reply examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java:50]: what do you mean with moving the configuration?

File: bucket4j-spring-boot-starter-context/pom.xml (hunk 1 above)
Comment (Edwin9292) [reply bucket4j-spring-boot-starter-context/pom.xml]: Looked at it again. It was added for the predicate validations. This however shouldn't have been tomcat, but jakarta.servlet with scope provided, good catch. It is now replaced.

File: examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java (hunk 2 above)
Comment (Edwin9292) [reply examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java:50]: Do you mean something like this?
https://github.com/Edwin9292/bucket4j-spring-boot-starter/pull/3 

Since the cacheManager is generic it's not possible to add such specific checks in that class. For this reason I made a Utils class for checking if the update is valid. The Jakarta validation should in my opinion still be done in the controller, same as the getting/updating of the cache. 

Please let me know if this pull request would resolve your comment or if you would like to see additional changes.

Comment (MarcGiffing) [reply examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java:50]: https://github.com/Edwin9292/bucket4j-spring-boot-starter/pull/3 is good. So the user must not implement the logic himself. Especially if there are changes in the logic. 

But I mean to provide a general update Controller out of the box. On solution would be a custom Spring Actuator Endpoint. It's only a proposal. It's ok as it is.
https://www.baeldung.com/spring-boot-actuators#custom-endpoint

Comment (Edwin9292) [reply examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/TestController.java:50]: Never worked with Spring Actuator before, but the example looks like it's an easy way to create endpoints as Beans. I can imagine this is a nice solution for users that don't wish (or know how) to configure the endpoint themself. 

However, I do see potential risks when adding such an endpoint in the background. Especially when people don't know that they have to configure Spring Security for that endpoint, since it could allow anyone to update configurations. Even if the endpoint could be enabled/disabled by some application property, I'm not sure if the company i'm creating the feature for would accept it. They like to have full control over which endpoints they have and what behavior they contain.

Since you agree with the provided example I will merge that branch. 
In case I do have spare time for implementing the other option that i mentioned, I'll create another pull request and send you a link for reviewing.

File: bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/Bucket4JConfigurationPredicateNameValidator.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import jakarta.validation.ConstraintValidatorContext;
+
+import org.springframework.http.server.reactive.ServerHttpRequest;
+
+import com.giffing.bucket4j.spring.boot.starter.context.ExecutePredicate;
+import com.giffing.bucket4j.spring.boot.starter.context.ExecutePredicateDefinition;
+import com.giffing.bucket4j.spring.boot.starter.context.FilterMethod;
+import com.giffing.bucket4j.spring.boot.starter.context.properties.Bucket4JConfiguration;
+
+public class Bucket4JConfigurationPredicateNameValidator implements ConstraintValidator<ValidPredicateNames, Bucket4JConfiguration> {
```
Comment (MarcGiffing) [inline bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/Bucket4JConfigurationPredicateNameValidator.java:21]: The problem here is, that the user is forced to add the the Servlet and Reactive dependencies to performa the validation. Or am I wrong? The jarkarta servlet dependency is also not scoped as provided so it is everytime added as a dependency. And thats only for the HttpServletRequest. The starter should work only with webflux or with the servlet api.

Comment (Edwin9292) [reply bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/Bucket4JConfigurationPredicateNameValidator.java:21]: It does require the Jakarta Servlet dependency yes, the Reactive was already included in the library through Spring itself. 
I did try to add the Jakarta dependency as scope provided, but that caused all kinds of exceptions saying the HttpServletRequest class could not be found. Maybe I did something wrong there, I could try it again tomorrow. Or do you perhaps know a fix for this?

Comment (MarcGiffing) [reply bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/Bucket4JConfigurationPredicateNameValidator.java:21]: Yes, the reactive ServerHttpRequest is in the spring-web dependency which will be used for Servlet and Reactive. But HttpServletRequest ist jakarta specific. I don't know how to fix it. With the Validator which uses both Request classes you're forced to add the dependency. I'll think about it again...

File: bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/DurationChronoUnitValidator.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import java.time.temporal.ChronoUnit;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+public class DurationChronoUnitValidator implements ConstraintValidator<ValidDurationChronoUnit, ChronoUnit> {
+
+	@Override
+	public boolean isValid(ChronoUnit value, ConstraintValidatorContext context) {
+		return value != null && (value == ChronoUnit.DAYS || !value.isDurationEstimated());
```
Comment (MarcGiffing) [inline bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/DurationChronoUnitValidator.java:12]: I'm trying to understand the check but I didn't get it. Can you please explain the part:
(value == ChronoUnit.DAYS || !value.isDurationEstimated()).

The method isDurationEstimated check if it equals DAYS:
return this.compareTo(DAYS) >= 0;

File: bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java [hunk 5]
```diff
@@ -3,34 +3,48 @@
...
 import com.giffing.bucket4j.spring.boot.starter.config.cache.SyncCacheResolver;
 import com.giffing.bucket4j.spring.boot.starter.context.properties.Bucket4JBootProperties;
+import com.giffing.bucket4j.spring.boot.starter.context.properties.Bucket4JConfiguration;
 
 @Configuration
 @ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
-@ConditionalOnClass({ Caching.class, JCacheCacheManager.class })
+@ConditionalOnClass({Caching.class, JCacheCacheManager.class})
 @ConditionalOnBean(CacheManager.class)
-@ConditionalOnMissingBean(SyncCacheResolver.class)
```
Comment (MarcGiffing) [inline bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java:22]: The reason for this check was to verify that this configuration is not applied when a different SynchCacheResolver is present to prevent two Configurations are activated at the same time. Is ist save to move it to the bean?

Comment (Edwin9292) [reply bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java:22]: This check is moved to line 32 above the method that creates the CacheResolver bean. This way the configuration class can also be used for creating the CacheManager and CacheListener beans, without being affected by an existing CacheResolver bean. The behaviour should still be the same, meaning that the CacheResolver bean will not be instantiated if one already exists.

File: bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/DurationChronoUnitValidator.java (hunk 4 above)
Comment (Edwin9292) [reply bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/DurationChronoUnitValidator.java:12]: The Duration.of() method which is used when creating rate limits, does not allow estimated timeunits. The 'DAYS' unit however is an exception to this rule. According to the "isDurationEstimated" method a day is an estimated unit because of daylight saving. The Duration.of() method on the other hand has a specific check that treats days as 24hours, making it no longer estimated. For this reason this specific check is added so Days are allowed. 

Good that you mention your confusion though, I will add a comment above the code to explain the behavior for future developers that look into it.

File: bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/Bucket4JConfigurationPredicateNameValidator.java (hunk 3 above)
Comment (Edwin9292) [reply bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/Bucket4JConfigurationPredicateNameValidator.java:21]: It took me quite some time to find a solution for this problem.. A simple @ConditionalOnClass would not work since it's not an @Configuration or @Service annotated class, causing the annotations to just be ingore. 

In the end I finally found a solution that does work. Can you let me know what you think of this solution before i merge it and mark the comment as resolved? 
https://github.com/Edwin9292/bucket4j-spring-boot-starter/pull/4

Comment (MarcGiffing) [reply bucket4j-spring-boot-starter-context/src/main/java/com/giffing/bucket4j/spring/boot/starter/context/constraintvalidations/Bucket4JConfigurationPredicateNameValidator.java:21]: 👍

File: bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java (hunk 5 above)
Comment (MarcGiffing) [reply bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java:22]: Isn't there a problem when to Caching Providers are in the classpath? For example JCache and Redis. If its the case, two of them are configured.

I long term solution would be to move each cache configuration to a separated Maven module, to make it explicit which cache configuration should be used and configured by this starter. But for now Its a problem. There was already someone in the past who has two cache implementation in the classpath with an configuration problem.

Comment (Edwin9292) [reply bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java:22]: The behaviour this should create is that the configuration class is always executed when (in case of JCache) a CacheManager bean exists, even if there already is a SyncCacheResolver. The method annotated with @Bean to create the JCacheCacheResolver does have the @ConditionalOnMissingBean annotation. This means that when someone creates his own instance of an AsyncCacheResolver, this bean will not created. The CacheManager and CacheListener for dynamically changing limits will however be made (if config caching is enabled).

Looking at it the way you said, where someone has multiple cache instances, it might be possible that 2 cache managers and listeners are created, So i can add the @ConditionalOnMissingBean annotation to those methods just to be sure. But the behaviour of creating the cache resolvers should still be functionally the same as how you implemented it.

Comment (MarcGiffing) [reply bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/jcache/JCacheBucket4jConfiguration.java:22]: OK 👍

**Comment by MarcGiffing** [comment]
Can you merge the current master into this branch?

**Comment by Edwin9292** [comment]
> Can you merge the current master into this branch?

Done, also added the missingbean annotation to the cachemanagers and added a comment to the ChronoUnit validator.

File: bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/redis/jedis/JedisCacheListener.java [hunk 6]
```diff
@@ -52,42 +46,13 @@ public JedisCacheListener(JedisPool jedisPool, String cacheName, Class<K> keyTyp
 	}
 
 	public void subscribe() {
-		Thread thread = new Thread(() -> {
```
Comment (MarcGiffing) [inline bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/redis/jedis/JedisCacheListener.java]: Was it intended to revert the changes from [https://github.com/Edwin9292/bucket4j-spring-boot-starter/pull/5](https://deref-web.de/mail/client/GzqxV05Trds/dereferrer/?redirectUrl=https%3A%2F%2Fgithub.com%2FEdwin9292%2Fbucket4j-spring-boot-starter%2Fpull%2F5) ?

Comment (Edwin9292) [reply bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/redis/jedis/JedisCacheListener.java]: No that was not intended at all. That's weird, looks like something went wrong with merging. Glad you caught it, will fix it right now

Comment (Edwin9292) [reply bucket4j-spring-boot-starter/src/main/java/com/giffing/bucket4j/spring/boot/starter/config/cache/redis/jedis/JedisCacheListener.java]: Looked into it, appears a revert before the merging had unexpected behavior that i missed. I reverted the revert, so it's fixed now. Marking this as resolved.
